clippy_lints/src/unused_io_amount.rs

```diff
@@ -0,0 +1,138 @@
+/// process the entire buffer. They return how many bytes were processed, which might be smaller
+/// than a given buffer's length. If you don't need to deal with partial-write/read, use
+/// `write_all`/`read_exact` instead.
+///
+/// **Known problems:** Detects only common patterns.
+///
+/// **Example:**
+/// ```rust,ignore
+/// use std::io;
+/// fn foo<W: io::Write>(w: &mut W) -> io::Result<()> {
+///     // must be `w.write_all(b"foo")?;`
+///     w.write(b"foo")?;
+///     Ok(())
+/// }
+/// ```
+declare_lint! {
+    pub UNUSED_IO_AMOUNT,
+    Deny,
+    "unused written/read amount"
+}
+
+pub struct UnusedIoAmount;
+
+impl LintPass for UnusedIoAmount {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(UNUSED_IO_AMOUNT)
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedIoAmount {
+    fn check_stmt(&mut self, cx: &LateContext, s: &hir::Stmt) {
+        let expr = match s.node {
+            hir::StmtSemi(ref expr, _) |
+            hir::StmtExpr(ref expr, _) => &**expr,
+            _ => return,
+        };
+
+        if let hir::ExprRet(..) = expr.node {
+            return;
+        }
+
+        match expr.node {
+            hir::ExprMatch(ref expr, ref arms, _) if is_try(arms) => {
+                if let hir::ExprCall(ref func, ref args) = expr.node {
+                    if let hir::ExprPath(ref path) = func.node {
+                        if match_path(path, &paths::CARRIER_TRANSLATE) && args.len() == 1 {
+                            check_method_call(cx, &args[0], expr);
+                        }
+                    }
+                } else {
+                    check_method_call(cx, expr, expr);
+                }
+            },
+
+            hir::ExprMethodCall(ref symbol, _, ref args) => {
+                let symbol = &*symbol.node.as_str();
+                match symbol {
+                    "expect" | "unwrap" | "unwrap_or" | "unwrap_or_else" => {
+                        check_method_call(cx, &args[0], expr);
+                    },
+                    _ => (),
+                }
+            },
+
+            _ => (),
+        }
+    }
+}
+
+fn check_method_call(cx: &LateContext, call: &hir::Expr, expr: &hir::Expr) {
+    if let hir::ExprMethodCall(ref symbol, _, _) = call.node {
+        let symbol = &*symbol.node.as_str();
+        if match_trait_method(cx, call, &paths::IO_READ) && symbol == "read" {
+            span_lint(cx,
+                      UNUSED_IO_AMOUNT,
+                      expr.span,
+                      "handle read amount returned or use `Read::read_exact` instead");
+        } else if match_trait_method(cx, call, &paths::IO_WRITE) && symbol == "write" {
+            span_lint(cx,
+                      UNUSED_IO_AMOUNT,
+                      expr.span,
+                      "handle written amount returned or use `Write::write_all` instead");
+        }
+    }
+}
+
+fn is_try(arms: &[hir::Arm]) -> bool {
```

Comment:
This looks useful enough that it could be in the `utils` mod :smile:.
(maybe `fn is_try(expr: &hir::Expr) -> Option<hir::Expr>` that returns `foo` in `foo?` if this is a `?` expression.)